Add initial content to the Sphinx Doc (#59)
alpaka-group/vikunja: docs/source/basic/operators.rst

Operators
=========

> The operator describes the operation applied to each element of a range of input elements. There are two ways to define an ``operator``: a lambda or a functor.

Reviewer: What about `std::function`, function pointers, ...?

Author: I didn't tested it. The main problem is CUDA. Some templates are not correctly resolved if there is a dependency between host code and a `__device__` function. That's the reason, why global `__device__` does not work. But to be hones, I don't know if functions pointers and `std::function` are working or not.

Author: `std::function` will not work with CUDA: https://nvidia.github.io/libcudacxx/standard_api/utility_library/functional.html

Reviewer: Okay, maybe also explain this here.

Author: I extend the warning box with a explanation.

        template<typename TData>
        ALPAKA_FN_HOST_ACC TData operator()(TData const val) const
        {
            return val + 1;
        }
    };

    int main(){
        // ...

        IncOne incOne;

        vikunja::transform::deviceTransform<Acc>(
            devAcc,
            queueAcc,
            extent[Dim::value - 1u],
            deviceNativePtr,
            deviceNativePtr, 
            incOne // Operator
    );

        return 0;
    }

The functor object must fullfil the requirements of ``std::is_trivially_copyable``. Compared to a lambda, the functor allows the creation of additional member variables and functions.

.. code-block:: c++

    struct MightFunctor {
    private:
        int const m_max;

        ALPAKA_FN_HOST_ACC int crop_value(int const v) const {
            return (v < m_max) ? v : m_max;
        }

    public:

        MightFunctor(int const max) : m_max(max) {}

        ALPAKA_FN_HOST_ACC TData operator()(TData const val) const
        {
            return crop_value(val);
        }
    }

    int main(){
        // ...

        MightFunctor mightFunctor(42);

        // ...

        return 0;

    }

.. warning:: 

>     Global functions are not allowed as functor objects for the vikunja ``algorithm`` due to a limitation of the Nvidia CUDA accelerator.

Reviewer: ```suggestion
    Global functions are not allowed as functor objects for the vikunja ``algorithm`` due to a limitation of the CUDA accelerator.
```

Operator Types
++++++++++++++

> Depending on the ``algorithm``, the ``operator`` requires a different number of input arguments. Currently, the vikunja ``algorithm`` requires a unary (one data input) or binary (two data inputs) ``operator``. A vikunja-specific property of the ``operator`` is that they can have an additional ``acc`` argument, which is required for some alpaka-specific functions.

Reviewer: ```suggestion
Depending on the ``algorithm``, the ``operator`` requires a different number of input arguments. Currently, the vikunja ``algorithm`` requires a unary (one data input) or binary (two data inputs) ``operator``. A vikunja-specific property of the ``operator`` is that they can have an additional ``acc`` argument, which is required for any device-specific functions (such as math functions).
```

.. code-block:: c++

    struct BinaryOperatorWithoutAccObject
    {
        template<typename TData>
        ALPAKA_FN_HOST_ACC TData operator()(TData const i, TData const j) const
        {
            return i + j;
        }
    };

    struct BinaryOperatorWithAccObject
    {
        template<typename TAcc, typename TData>
        ALPAKA_FN_HOST_ACC TData operator()(TAcc const& acc, TData const i, TData const j) const
        {
            return alpaka::math::max(acc, i, j);
        }

Please read the `alpaka documentation <https://alpaka.readthedocs.io/en/latest/index.html>`_ for more information about available device functions.

.. warning:: 

>     The ``acc`` object also allows access to functions that can break the functionality of the vikunja ``algorithm``, such as using the thread index for a manual memory access.

Reviewer: I wonder whether it would make sense to provide a `vikunja::acc` type that capsules the alpaka accelerator and only provides access to the safe functions.

Author: There are pro and cons for this idea.

(+) It makes safer to use the code.

(-) It requires extra code to maintain.
(-) Some function are malicious depending on context but allows also new functionality. For example the thread id. If you use it for memory access and read data, most of the time it is no problem and allows for example stencil code. But if you use it for write access, it could break the algorithm.

Reviewer: Is a stencil code something that should be supported by the current algorithms? How would this work with `std::transform` or `std::reduce`?

Author: It is possible with `std::transform` and a counting iterator. In PIConGPU, stencil codes are calculated with a parallel `for_each` and the thread id. I opened an issue for discussion https://github.com/alpaka-group/vikunja/issues/67 . At the moment I'm not sure if it is really useful. But I can say, it has no high priority. Therefore an issue is a good way to not forget it.
